aws/s3.go

```diff
@@ -56,87 +65,197 @@ func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error
-			logging.Logger.Debugf("Creating session - Bucket %s - region %s", *bucket.Name, bucketRegion)
-			awsSession, err := session.NewSession(&awsgo.Config{
-				Region: awsgo.String(bucketRegion)},
-			)
-			if err != nil {
-				return bucketNamesPerRegion, errors.WithStackTrace(err)
-			}
-			regionClients[bucketRegion] = s3.New(awsSession)
+// getBucketNamesPerRegions gets valid bucket names concurrently from list of target buckets
+func getBucketNamesPerRegion(svc *s3.S3, targetBuckets []*s3.Bucket, excludeAfter time.Time, regionClients map[string]*s3.S3,
+	bucketNameSubStr string) (map[string][]*string, error) {
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+	var bucketCh = make(chan *S3Bucket, len(targetBuckets))
+	var wg sync.WaitGroup
+
+	for _, bucket := range targetBuckets {
+		if len(bucketNameSubStr) > 0 && !strings.Contains(*bucket.Name, bucketNameSubStr) {
+			logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
+			continue
 		}
 
-		validTags, excludeReason := hasValidTags(regionClients[bucketRegion], bucket)
-		if !validTags {
-			logging.Logger.Debugf("Skipping - Bucket %s - %s", *bucket.Name, excludeReason)
+		wg.Add(1)
+		go func(bucket *s3.Bucket) {
+			defer wg.Done()
+			getBucketInfo(svc, bucket, excludeAfter, regionClients, bucketCh)
+		}(bucket)
+	}
+
+	go func() {
+		wg.Wait()
+		close(bucketCh)
+	}()
+
+	for bucketData := range bucketCh {
+		if bucketData.Error != nil {
+			logging.Logger.Warnf("Skipping - Bucket %s - region - %s - error: %s", bucketData.Name, bucketData.Region, bucketData.Error)
+			continue
+		}
+		if !bucketData.IsValid {
+			logging.Logger.Debugf("Skipping - Bucket %s - region - %s - %s", bucketData.Name, bucketData.Region, bucketData.InvalidReason)
 			continue
 		}
+		if _, ok := bucketNamesPerRegion[bucketData.Region]; !ok {
+			bucketNamesPerRegion[bucketData.Region] = []*string{}
+		}
+		bucketNamesPerRegion[bucketData.Region] = append(bucketNamesPerRegion[bucketData.Region], &bucketData.Name)
+	}
+	return bucketNamesPerRegion, nil
+}
 
-		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
-			bucketNamesPerRegion[bucketRegion] = []*string{}
+// getBucketInfo populates the local S3Bucket struct for the passed AWS bucket
+func getBucketInfo(svc *s3.S3, bucket *s3.Bucket, excludeAfter time.Time, regionClients map[string]*s3.S3, bucketCh chan<- *S3Bucket) {
+	var bucketData S3Bucket
+	bucketData.Name = *bucket.Name
+	bucketData.CreationDate = *bucket.CreationDate
+
+	bucketRegion, err := getS3BucketRegion(svc, bucketData.Name)
+	if err != nil {
+		bucketData.Error = err
+		bucketCh <- &bucketData
+		return
+	}
+	bucketData.Region = bucketRegion
+
+	// Check if the bucket is in target region
+	matchedRegion := false
+	for region := range regionClients {
+		if region == bucketData.Region {
+			matchedRegion = true
+			break
 		}
+	}
+	if !matchedRegion {
+		bucketData.InvalidReason = "Not in target region"
```

Comment:
How would we have gotten a bucket not in one of those regions? Is this just an extra safety check?

Reply:
For S3 - you have to list out all buckets and find each bucket's individual region. Earlier we were listing all buckets, finding all regions and returning it. This wasn't required because if there is a target region specified by end user (or all regions if not specified) - and if a bucket is not in the target region - we will be skipping it later - better to skip it at the start.